src/types.ts

```diff
@@ -904,14 +907,16 @@ export class Signature {
     program: Program,
     parameterTypes: Type[] | null = null,
     returnType: Type | null = null,
-    thisType: Type | null = null
+    thisType: Type | null = null,
+    /** Number of required parameters excluding `this`. Other parameters are considered optional. */
+    public readonly requiredParameters: i32 = parameterTypes ? parameterTypes.length : 0,
+    /** Whether the last parameter is a rest parameter. */
+    public readonly hasRest: bool = false,
```

Comment:
Suggesting to unify these to also use explicit fields to fit in with the other parameters. Alternatively, the old ones could also be changed to become constructor parameters like the new ones, if that's possible. Either way is fine for me as long as code becomes uniform.

Reply:
These properties are now constructor parameters.